### ISSUE TITLE
Allow to use single control flow statements into expression bodied members

### ISSUE THREAD

**Issue by Makeloft** [dotnet/csharplang#1179]
Allow using of single control flow statements [operators: switch, if, else, yield, using, foreach, for, do, while] like lambda expresions to avoid brackets

        public static void SetValue(this MemberInfo member, object obj, object value)
        {
            switch (member)
            {
                case PropertyInfo property:
                    property.SetValue(obj, value, null);
                    break;
                case FieldInfo field:
                    field.SetValue(obj, value);
                    break;
            }
        }

        public static void SetValue(this MemberInfo member, object obj, object value) =>
            switch (member)
            {
                case PropertyInfo property:
                    property.SetValue(obj, value, null);
                    break;
                case FieldInfo field:
                    field.SetValue(obj, value);
                    break;
            }

        public static void SetValue(this MemberInfo member, object obj, object value) => switch (member)
        {
            case PropertyInfo property:
                property.SetValue(obj, value, null);
                break;
            case FieldInfo field:
                field.SetValue(obj, value);
                break;
        }

        public static void ForEach<T>(this IEnumerable<T> collection, Action<T> action)
        {
            foreach (var item in collection) action(item);
        }

        public static void ForEach<T>(this IEnumerable<T> collection, Action<T> action) =>
            foreach (var item in collection) action(item);

        public static IEnumerable<T> ToEnumerable<T>(this T singleItem)
        {
            yield return singleItem;
        }

        public static IEnumerable<T> ToEnumerable<T>(this T singleItem) =>
            yield return singleItem;

**Comment by MkazemAkhgary** [comment]
One side note. You dont have to write every thing including every single example in title. Title is just a  name of posting and should only describe what this post is all about. You write rest of things in posting.

Title could be: allow to use "control flow statements" in expression bodied members.

**Comment by theunrepentantgeek** [comment]
I don't think `{` braces `}` are an enemy that needs to be vanquished.

Expression bodied members are a great addition to the language, no doubt. 

My experience with them has convinced me that they're best used for simple expressions, with an upper bound on complexity. As a personal choice, I revert to using braces as soon as the expression extends past a single line - I find the enforced whitespace and explicit delimiters help maintain the readability of the code.

I'm not sure there's value in making language changes to accommodate this edge case.

**Comment by Makeloft** [comment]
Maybe for some operators like if-else or some cycles brackets makes reading better but I sure for switch, yeild, using brackets may be avoided.

Really, I love C# 7 mostly for functional style which allows to write single or two lined methods. :)

**Comment by aluanhaddad** [comment]
One we have `match` expressions, I think this will be implicitly resolved in a much more principled and coherent manner.

**Comment by Makeloft** [comment]
@aluanhaddad is there public discussion about `match` operator ?

**Comment by DavidArno** [comment]
@theunrepentantgeek,

I do not see "must be on a single line" as a wise choice of rule around when to use/not use expression-bodied members. For example, in my view, the following is a perfectly valid use of this feature:
```cs
private ConsEnumerable(IEnumerable<T> enumeration, IEnumerable<T> head) =>
    _node = new ConsNode<T>
    {
        State = StartNode,
        Next = new ConsNode<T>
        {
            Enumeration = head,
            State = HasEnumeration,
            Next = EnumerationConsNode(enumeration)
        }
    };
```
Following the "all on one line" rule would make turn such an initialiser into a long, messy, single line. Yet adding `{}` and a `return` just adds pointless noise to this piece of code, without improving its readability.

> I'm not sure there's value in making language changes to accommodate this edge case.

On what basis do you conclude this is an edge case? I think there are plenty of folk that would use expression-bodied syntax for single statement methods. However, whilst it's not edge case, I still think this would be a mistake. For example, why add support for:
```cs
public static IEnumerable<T> ToEnumerable<T>(this T singleItem) => yield return singleItem;
```
when the following is sufficient?
```cs
public static IEnumerable<T> ToEnumerable<T>(this T singleItem) => new[] {singleItem};
```
and I'm not convinced that *expression*-bodied methods are the best choice for statements like `foreach`, `switch` etc.

**Comment by IanWold** [comment]
*tl;dr this proposal increases my own readability for the switch statement example provided, but I have some (potentially unfounded) reservations regarding more complex possibilities.*

I would advocate in favor of this proposal (or a flavor of it) due to the following example OP provided:

```c#
    public static void SetValue(this MemberInfo member, object obj, object value) => switch (member)
    {
        case PropertyInfo property:
            property.SetValue(obj, value, null);
            break;
        case FieldInfo field:
            field.SetValue(obj, value);
            break;
    }
```

There are two reasons this proposed syntax is more meaningful to me than the current syntax. First, the concern regarding any added reading complexity due to the elimination of curly braces does not really apply to this case, as the switch statement already has curly braces around the case statements. Second, the purpose of this method is just to switch on its input, and the proposed syntax conveys that to me better than the current syntax. As a result, I personally find that this proposal enhances clarity for the above case.

However, I am worried about the following possibility:

```c#
public static void Something(bool someBool) =>
    if (someBool) doSomething();
    else doSomethingElse();
```

Though my contrived example is easy to read, it's easy to imagine a larger if statement (or a series of nested conditionals and loops) could become very unreadable. But they already exist and they're already unreadable, so I'm unsure if that's a concern which I should have specifically in regards to this proposal?

**Comment by DavidArno** [comment]
@Makeloft,

> @aluanhaddad is there public discussion about `match` operator ?

See #1054.

**Comment by HaloFour** [comment]
@DavidArno 

Personally (and I recognize that this is strictly stylistic preference), I find the use of expression-bodied members in your example to be less readable than the imperative version with braces.  The difference in verbosity is minimal.  Allman adds 2 LOC (I prefer K&R).  My preference is to use expression-bodied members when the body is an expression.  I apply the same guidelines to them as I would any other expressions.  Object initializers would be a perfectly cromulent use-case, in my opinion.

I don't see the necessity of expanding expression-bodied members, at least in the scenarios listed in this proposal.  I think `match` (or whatever form the expression `switch` takes) will solve numerous other potential use cases.  What really concerns me is what is likely to happen when some developers start to push the envelope on mixing sequence expressions and expression-bodied members.  Without judicious limits I fear that a really awkward and really ugly dialect of C# will start brewing, all in the name of trying to make C# into Python.

**Comment by Makeloft** [comment]
@DavidArno I don't sure but looks like `new[] {singleItem};` will not so effective with memory usage then `yield return singleItem;` if we call ToEnumerable method many-many times.

**Comment by HaloFour** [comment]
@Makeloft 

An iterator method also causes allocations as the compiler has to lift the method into a class that implements `IEnumerable<T>` and `IEnumerator<T>`.  If you want to do that I would suggest creating an array or other singleton enumerator upfront and always return that.

**Comment by DavidArno** [comment]
@Makeloft,

Compare what the compiler does [when using an array](https://sharplab.io/#v2:EYLgZgpghgLgrgJwgZwLRKgEwPYDsA2AnuhGAD4ACADAAQUCMA3ALABQ1d9ALC62xQGZOANjoAmGgGEabAN5saiukIaiGAgDwAVAHw0t2AKK44AWwgIowfBG06AFDAAWAS2T6ayF7gDmNgJIwEKYAlDQAvHq4EADuANoAujSyXr4BQaYAvryZbEA) versus [using an iterator](https://sharplab.io/#v2:EYLgZgpghgLgrgJwgZwLRKgEwPYDsA2AnuhGAD4ACADAAQUCMA3ALABQ1d9ALC62xQGZOANjoAmGgGEabAN5saiukIaiGAgDwAVAHw0t2AKK44AWwgIowfBDHadAChgALAJbJ9NZK9wBzGwCSMBCmAJQyrEo08pFRSgz0dADsXj7+EEEhvFEAvmx5rEA). You might be in for a shock over what that simple `yield return` gets lowered too! 😀

**Comment by Makeloft** [comment]
@DavidArno as I understand when we iterating by array iterator created too.
When we use `yield return singleItem;` we create only iterator instance
When we use `new[] {singleItem};` we create array into method ToEnumerate and then when iterate by collection create iterator. Maybe array implementation contains iterator logic and methods, I don't know, but in this way array instance creation require more resources.

 Also interesting to see memory-weak safe singleton implementation of this method ToEnumerable (without WeakReferences ;) )

**Comment by Makeloft** [comment]
@HaloFour 
C# is multiparadigmatic language. It is one of strongest advantages.
And every developer may wrote code into individual style. If you like braces you may wrote them, but I like short one or two lined methods. Now I can wrote them but not in all possible cases.

**Comment by DavidArno** [comment]
@HaloFour,
> Object initializers would be a perfectly cromulent use-case, in my opinion.

Yet the language allows them and allows `=>` to be used for `void` methods too. I use expression-bodied members for variable assignments and expressions, which (as far as I'm aware) is what they were designed for,

**Comment by HaloFour** [comment]
@Makeloft 

> C# is multiparadigmatic language. It is one of strongest advantages.

From a functionality point of view, yes.  Not from a grammatical point of view.  C# isn't going to adopt the syntax from other programming languages simply because it satisfies some stylistic preferences of programmers who wish they were actually using those other languages.  The CLR supports a wide range of languages with different styles.  If you want those styles, feel free to pick one.  C# will never become F#, or Python, or Swift, or whatever.  It may adopt worthwhile features from those languages where they make sense, but the syntax will be adapted to fit with C#.

**Comment by Makeloft** [comment]
Sure that C# can contains some of best features of another languages adapted into realities (including grammar). :)

**Comment by Makeloft** [comment]
Do anybody really interested on this feature or I can close it ?

**Comment by gafter** [comment]
@Makeloft You're welcome to close this.
